### PR TITLE
Authorization rule changes for IDPs isolation [DPP-1336]

### PR DESCRIPTION
**Pull request by skisel-da**

### CODE REVIEW

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/IdentityProviderAwareAuthService.scala [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+class IdentityProviderAwareAuthService(
+    defaultAuthService: AuthService,
+    configLoader: ConfigLoader,
+    jwtVerifierLoader: JwtVerifierLoader,
+)(implicit
+    executionContext: ExecutionContext,
+    loggingContext: LoggingContext,
+) extends AuthService {
+
+  private implicit val logger: ContextualizedLogger = ContextualizedLogger.get(this.getClass)
```
Comment (pbatko-da) [inline ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/IdentityProviderAwareAuthService.scala:27]:
```suggestion
  private implicit val logger: ContextualizedLogger = ContextualizedLogger.get(getClass)
```

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/IdentityProviderAwareAuthService.scala [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+          ClaimSet.Unauthenticated
+        }
+    }).asJava
+      .thenCompose(defaultAuth(headers))
+  }
+
+  def defaultAuth(
+      headers: Metadata
+  )(prevClaims: ClaimSet): CompletionStage[ClaimSet] =
+    if (prevClaims != ClaimSet.Unauthenticated)
```
Comment (pbatko-da) [inline ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/IdentityProviderAwareAuthService.scala:44]: What does this code do?

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/CachedJwtVerifierLoader.scala [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+object CachedJwtVerifierLoader {
+
+  case class CacheKey(
+      jwksUrl: JwksUrl,
+      keyId: Option[String],
+  )
+
+  case class CacheConfig(
+      // Large enough such that malicious users can't cycle through all keys from reasonably sized JWKS,
```
Comment (pbatko-da) [inline ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/CachedJwtVerifierLoader.scala:77]: What does it mean that a large enough cache prevent's cycling through all the keys?

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/CachedJwtVerifierLoader.scala [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      // forcing cache eviction and thus introducing additional latency.
+      maxSize: Long = 1000,
+      expirationTime: Long = 10,
+      expirationUnit: TimeUnit = TimeUnit.HOURS,
```
Comment (pbatko-da) [inline ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/CachedJwtVerifierLoader.scala:81]: Is `hours` deliberate?

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/CachedJwtVerifierLoader.scala [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  )
+
+  case class HttpConfig(
+      connectionTimeout: Long = 10,
```
Comment (pbatko-da) [inline ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/CachedJwtVerifierLoader.scala:85]: If these configs values are not configurable, then maybe use `objects` instead of `case class`es?

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Claims.scala [hunk 6]
```diff
@@ -118,6 +121,14 @@ object ClaimSet {
     def isAdmin: Either[AuthorizationError, Unit] =
       Either.cond(claims.contains(ClaimAdmin), (), AuthorizationError.MissingAdminClaim)
 
+    /** Returns true if the set of claims authorizes the user to use admin services or Identity Provider admin services, unless the claims expired */
+    def isAdminOrIDPAdmin: Either[AuthorizationError, Unit] =
+      Either.cond(
+        claims.contains(ClaimIdentityProviderAdmin) || claims.contains(ClaimAdmin),
```
Comment (pbatko-da) [inline ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Claims.scala:127]: Idea: You could represent both `ClaimIdentityProviderAdmin` and `ClaimAdmin` as:
```
case class ClaimAdminOver(canManageIdp: IdpRange)

trait IdpRange
case object All extends IdpRange
case class Idp(idpId) extends IdpRange
```

This might be beneficial if you verifying idps is needed only for Admins/IdpAdmins

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/services/UserManagementServiceAuthorization.scala [hunk 7]
```diff
@@ -29,39 +29,60 @@ private[daml] final class UserManagementServiceAuthorization(
     new DamlContextualizedErrorLogger(logger, loggingContext, None)
 
   override def createUser(request: CreateUserRequest): Future[CreateUserResponse] =
-    authorizer.requireAdminClaims(service.createUser)(request)
+    request.user match {
+      case Some(user) =>
+        authorizer.requireAdminIDPContext(user.identityProviderId, service.createUser)(request)
+      case None =>
+        authorizer.requireAdminOrIDPAdminClaims(service.createUser)(request)
```
Comment (pbatko-da) [inline ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/services/UserManagementServiceAuthorization.scala:36]: Could we treat `user = None` the same as `user.identityProviderId = ""`?
That would aslo match grpc's default value semantics

File: ledger/metrics/src/main/scala/com/daml/metrics/IdentityProviderConfigStoreMetrics.scala [hunk 8]
```diff
@@ -17,6 +17,7 @@ class IdentityProviderConfigStoreMetrics(
 ) extends FactoryWithDBMetrics {
 
   val cacheByIssuer = new CacheMetrics(prefix :+ "cache_by_issuer", registry)
```
Comment (pbatko-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/IdentityProviderConfigStoreMetrics.scala:19]:
```suggestion
  val idpConfigCache = new CacheMetrics(prefix :+ "idp_config_cache", registry)
```
?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala [hunk 9]
```diff
@@ -146,26 +148,25 @@ private[apiserver] final class ApiPartyManagementService private (
...
           .zip(partyRecords)
-          .collect(relevantParties(identityProviderId))
+          .map(disclosedPartyInformation(identityProviderId))
         ListKnownPartiesResponse(protoDetails)
       }
     }
   }
 
-  private def relevantParties(
+  private def disclosedPartyInformation(
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala:157]:
```suggestion
  private def blindAndConvertToProto(
```
?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala [hunk 10]
```diff
@@ -362,6 +367,22 @@ private[apiserver] final class ApiPartyManagementService private (
     }
   }
 
+  // Here we check if party exists and actually belongs to the requested Identity Provider
+  private def partyBelongsToIDPOrError(
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala:371]:
```suggestion
  private def verifyPartyIsNonExistentOrInIdp(
```
?
Current name suggests that a non-existent party, which obviously doesn't belong to any idp, would case an error

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala [hunk 11]
```diff
@@ -362,6 +367,22 @@ private[apiserver] final class ApiPartyManagementService private (
...
+      identityProviderId: IdentityProviderId,
+      party: Ref.Party,
+  )(implicit errorLogger: DamlContextualizedErrorLogger): Future[Unit] = {
+    partyRecordStore.getPartyRecordO(party).flatMap {
+      case Right(Some(party)) if party.identityProviderId != identityProviderId =>
+        Future.failed(
+          LedgerApiErrors.AuthorizationChecks.PermissionDenied
+            .Reject(s"Party ${party} belongs to another Identity Provider")
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala:379]:
```suggestion
            .Reject(s"Party ${party} belongs to an identity provider that differs from the one specified in the request")
```
?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala [hunk 12]
```diff
@@ -100,6 +100,7 @@ private[apiserver] final class ApiUserManagementService(
       } { case (user, pRights) =>
         for {
           _ <- identityProviderExistsOrError(user.identityProviderId)
+          _ <- userBelongsToIDPOrError(user.identityProviderId, user.id)
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala:103]: This probably doesn't make sense for the create user rpc

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala [hunk 13]
```diff
@@ -373,6 +362,24 @@ private[apiserver] final class ApiUserManagementService(
           )
       }
 
+  // Here we check if user exists and actually belongs to the requested Identity Provider
+  // Alternatively, identity_provider_id could be part of the compound unique key within user database.
```
Comment (pbatko-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala:366]: There's probably no need to make `identity_provider_id` part of a database key to have this check implemented in the persistence layer. Implementing it there would have two benefits:
- the check would be done within a db transaction,
- we would avoid doing the additional `getUser` calls in this method.

File: ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/api/IdentityProviderConfigStore.scala [hunk 14]
```diff
@@ -39,6 +39,20 @@ trait IdentityProviderConfigStore {
...
 
+  final def getActiveIdentityProviderByIssuer(issuer: String)(implicit
+      loggingContext: LoggingContext
+  ): Future[IdentityProviderConfig] =
+    getIdentityProviderConfig(issuer)
+      .flatMap {
+        case Right(value) if !value.isDeactivated => Future.successful(value)
+        case Right(value) =>
+          Future.failed(
+            new Exception(s"Identity Provider ${value.identityProviderId.value} is deactivated.")
```
Comment (pbatko-da) [inline ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/api/IdentityProviderConfigStore.scala:50]: Use error codes?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala (hunk 12 above)
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala:103]: its this funny case which we discussed already, if there is a user exists with a different IDP assigned - it should throw a permission denied

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Authorizer.scala [hunk 15]
```diff
@@ -82,6 +82,55 @@ final class Authorizer(
...
+
+  private def validateRequestIdentityProviderId(
+      requestIdentityProviderId: Option[IdentityProviderId],
+      claims: ClaimSet.Claims,
+  ): Either[AuthorizationError, Unit] = claims.identityProviderId match {
+    case id: IdentityProviderId.Id if !requestIdentityProviderId.contains(id) =>
+      // Claim is valid only for the specific Identity Provider,
+      // and identity_provider_id in the request matches the one provided in the claim.
+      Left(AuthorizationError.InvalidIdentityProviderId(id))
+    case _ =>
```
[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Authorizer.scala:130]: If `requestIdentityProviderId` is not provided - it's NOT an authorization problem, it will be validated out in the layer below.  It is an invalid argument which will be validated further. Not sure how to act here to make sure proper layers are responsible for errors.

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/CachedJwtVerifierLoader.scala (hunk 3 above)
Comment (skisel-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/CachedJwtVerifierLoader.scala:77]: More info here actually: https://digitalasset.atlassian.net/browse/DPP-681

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/CachedJwtVerifierLoader.scala (hunk 4 above)
Comment (skisel-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/CachedJwtVerifierLoader.scala:81]: This is what we use for normal JWKS authorization. I keep here the same default

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/IdentityProviderAwareAuthService.scala (hunk 2 above)
Comment (skisel-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/IdentityProviderAwareAuthService.scala:44]: it is `orElse`-like concept for Java's `CompletionStage`. I.e. we check for the authorization IDPs and if they're giving "access denied" - we fallback to default authorization config

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/services/UserManagementServiceAuthorization.scala (hunk 7 above)
Comment (skisel-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/services/UserManagementServiceAuthorization.scala:36]: We discussed this with @mziolekda and agreed that we do not have this convenience for `identity_provider_id` argument.

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/IdentityProviderAwareAuthService.scala (hunk 2 above)
Comment (pbatko-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/IdentityProviderAwareAuthService.scala:44]: Ok, so `prevClaims` are the claims resolved using the IDP aware auth service.

I'm guessing the fallback is needed because default IDP config (from the static configuration) is handled separately from the IDP configs stored in the index-db. Would it be beneficial to handle both together? E.g. by representing the default IDP as `com.daml.ledger.api.domain.IdentityProviderConfig` ?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala (hunk 13 above)
Comment (pbatko-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala:366]: Additionally this pretty much would create per idp username namespaces in turn would lead to cleaner errors in cases like `creating a user 'alice' in idp 'foo' when an 'alice' already exists in 'bar'`

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/IdentityProviderAwareAuthService.scala (hunk 2 above)
Comment (skisel-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/IdentityProviderAwareAuthService.scala:44]: Default authorization might not be necessary JWKS-based, but any other config option, so it cannot be handled by this code here.

Comment (skisel-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/IdentityProviderAwareAuthService.scala:44]: Also Canton has it's own layer on top

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Authorizer.scala (hunk 15 above)
[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Authorizer.scala:130]: Let me first define here what is what.

`claims.identityProviderId` is what is provided from the token. So two options here are possible:
1. default IDP (token is issued by the auth config defined in the default configuration) 
2. non-default IDP having concrete value (as it is of type `IdentityProviderId`)

`requestIdentityProviderId` is what we extracted out from the request. Here we distinguish the following cases:
1. `None`, means `identity_provider_id` from the request which cannot be recognized/parsed. For example String with more than 255 chars, or having invalid characters. 
2. Some value of non-default IDP which is equal to the IDP-ID of the token
3. Some value of non-default IDP which is NOT equal to the IDP-ID of the token
4. Default IDP (i.e. `identity_provider_id` was an empty string)

What is important to know with the authorization rules we have introduced:
1. A `participant_admin` may provide any arbitrary value as IDP-ID (i.e. `requestIdentityProviderId` does not matter) and authorization layer should not be restricting the admin (that is why if `claims.identityProviderId == Default` it is giving `Right(())`
2. An invalid `IdentityProviderId`, i.e. non-`LedgerString`, would give `None` for `requestIdentityProviderId`  and authorization layer would allow to pass the filter (but currently only for non-Default IDP), as this is nothing to do with `PERMISSION_DENIED` error which can be issued here.

You're right with your statement that `None` gives `InvalidIdentityProviderId`.  So `PERMISSION_DENIED` will be issued in the case of non-default IDP.  I think I just have to fix it and pass it further to the business logic layer. 

I think in general this highlights a bit broken design of the authorization layer. I don't like that in case of invalid request I'm passing things further and let it fail on the business logic. But doing it differently would break current pattern and will be much more work. Do you maybe have suggestions to improve that?

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Authorizer.scala [hunk 16]
```diff
@@ -82,6 +82,55 @@ final class Authorizer(
...
 
+  def requireAdminIDPContext[Req, Res](
+      identityProviderId: String,
+      call: Req => Future[Res],
+  ): Req => Future[Res] =
+    authorize(call) { claims =>
+      for {
+        _ <- valid(claims)
+        _ <- claims.isAdminOrIDPAdmin
+        requestIdentityProviderId = IdentityProviderId.fromString(identityProviderId).toOption
```
[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Authorizer.scala:93]: see my comment above re lines 127-130

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto [hunk 17]
```diff
@@ -198,6 +198,9 @@ message PartyDetails {
   ObjectMeta local_metadata = 4;
 
   // The id of the ``Identity Provider``
-  // Optional, if not set, assume the party is managed by the default identity provider or party is not hosted by the participant.
+  // Optional, if not set, there could be 3 options:
+  // 1) the party is managed by the default identity provider.
+  // 2) party is not hosted by the participant.
+  // 3) party is hosted by the participant, but is outside of the user's identity provider.
   string identity_provider_id = 5;
```
[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto:205]: > This might be a radical change at this point but if we were to represent the default idp as a non-empty string (e.g. "default") then we could simplify this to say 2) and 3) only (i.e. we wouldn't have to mention 1) separately)

Agreed offline we wont be doing it.

> Also, if you're the an admin in the default idp you won't be able to tell whether a party is yours or not because in all cases this field is set to the empty string, correct?

Correct, but not sure if this is what we want. `participant_admin` may lookup all known IDPs to this participant and get parties by which IDP then. This way it would be possible for him to learn all "participant-local" IDPs.

> Does bool is_local = 3;'s description needs an update to to say what happens if the party is hosted by in a different idp?

Will add a comment there too

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala (hunk 13 above)
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala:366]: As agreed offline - we dont do it

File: ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/api/IdentityProviderConfigStore.scala (hunk 14 above)
Comment (skisel-da) [reply ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/api/IdentityProviderConfigStore.scala:50]: Actually, I have a question here :) This exception is not thrown to the gRPC layer, to the user - this is covered by the caller code which is parsing a token and gives result. In the end nothing is thrown to the user, do you still think we need to throw an error code here?

**Comment by skisel-da** [comment]
> What if the user management service is enabled and someone uses a custom daml access token (one doesn't mention the user id). Can such a request succeed?

This should work as before. From the IDP perspective it would be as if request is coming from the default IDP.

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/CachedJwtVerifierLoader.scala (hunk 3 above)
Comment (skisel-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/CachedJwtVerifierLoader.scala:77]: It just has to be much bigger than 1-2 * 16 (max jwks urls). I suspect it is still bigger than 1000 :)

File: ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/api/IdentityProviderConfigStore.scala (hunk 14 above)
Comment (pbatko-da) [reply ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/api/IdentityProviderConfigStore.scala:50]: Some error codes appear just in the participant server logs and are useful to participant operators, so by default I'd use error codes.

Are you sure this won't be transmitted to the client? Who catches/handles it then?

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto [hunk 18]
```diff
@@ -188,7 +188,7 @@ message PartyDetails {
   // Optional
   string display_name = 2;
 
-  // true if party is hosted by the participant.
+  // true if party is hosted by the participant and party is within the user's identity provider.
```
Comment (pbatko-da) [inline ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto:191]: `within` means different things if you are an admin vs. non-admin. Can expand on that?

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Claims.scala [hunk 19]
```diff
@@ -118,6 +121,14 @@ object ClaimSet {
     def isAdmin: Either[AuthorizationError, Unit] =
       Either.cond(claims.contains(ClaimAdmin), (), AuthorizationError.MissingAdminClaim)
 
+    /** Returns true if the set of claims authorizes the user to use admin services or Identity Provider admin services, unless the claims expired */
```
Comment (pbatko-da) [inline ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Claims.scala:124]:
```suggestion
    /** Returns true if the set of claims authorizes the user as an administrator or an identity provider administrator, unless the claims expired */
```

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/ConfigLoader.scala [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.api.auth
+
+import com.daml.ledger.api.domain.IdentityProviderConfig
+import com.daml.logging.LoggingContext
+
+import scala.concurrent.Future
+
+trait ConfigLoader {
```
Comment (pbatko-da) [inline ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/ConfigLoader.scala:11]:
```suggestion
trait IdentityProviderConfigLoader {
```
?

**Review by pbatko-da — APPROVED** [review]
Looks good. Are you planning to add any tests to this PR?

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto (hunk 18 above)
Comment (skisel-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto:191]: it is actually not different. a participant_admin will see all parties which are not assigned to default idp as non-local

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Claims.scala [hunk 21]
```diff
@@ -167,6 +179,7 @@ object ClaimSet {
       applicationId = None,
       expiration = None,
       resolvedFromUser = false,
+      identityProviderId = IdentityProviderId.Default,
```
[the remark this answers is not in the document]
Comment (skisel-da) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Claims.scala:182]: It only holds the most authority for the `ParticipantAdmin`, which is intended. WIthout the role, user cannot do anything within this IDP.

File: ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/api/IdentityProviderConfigStore.scala (hunk 14 above)
Comment (skisel-da) [reply ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/api/IdentityProviderConfigStore.scala:53]: i have added a comment there

Comment (skisel-da) [reply ledger/participant-local-store/src/main/scala/com/daml/platform/localstore/api/IdentityProviderConfigStore.scala:53]: Forgot to mention as well for others - we spoke about this case offline

File: ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto (hunk 18 above)
Comment (skisel-da) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto:191]: @mziolekda I would appreciate your help on documenting this one...

Comment (mziolekda) [reply ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/party_management_service.proto:191]: Maybe rephrase to say:
```suggestion
  // true if party is hosted by the participant and the party shares the same identity provider as the user issuing the request.
```

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/AuthorizationError.scala [hunk 22]
```diff
@@ -53,4 +55,13 @@ object AuthorizationError {
   final case class MissingActClaim(party: String) extends AuthorizationError {
     override val reason = s"Claims do not authorize to act as party '$party'"
   }
+
+  final case class InvalidIdentityProviderId(identityProviderId: IdentityProviderId)
+      extends AuthorizationError {
+    private val id = identityProviderId.toRequestString
+    override val reason =
+      s"identity_provider_id from the request `$id` does not match the one provided in the authorization claims'."
```
Comment (mziolekda) [inline ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/AuthorizationError.scala:63]:
```suggestion
      s"identity_provider_id from the request `$id` does not match the one provided in the authorization claims"
```

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala [hunk 23]
```diff
@@ -100,6 +100,7 @@ private[apiserver] final class ApiUserManagementService(
       } { case (user, pRights) =>
         for {
           _ <- identityProviderExistsOrError(user.identityProviderId)
+          _ <- verifyUserBelongsToIDPOrError(user.identityProviderId, user.id)
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala:103]: How does that work? We are checking in the db if the user belongs to the idp before we actually add it to the db?

File: ledger/metrics/src/main/scala/com/daml/metrics/IdentityProviderConfigStoreMetrics.scala [hunk 24]
```diff
@@ -16,7 +16,8 @@ class IdentityProviderConfigStoreMetrics(
     override val registry: MetricRegistry,
 ) extends FactoryWithDBMetrics {
 
-  val cacheByIssuer = new CacheMetrics(prefix :+ "cache_by_issuer", registry)
+  val idpConfigCache = new CacheMetrics(prefix :+ "idp_config_cache", registry)
```
Comment (mziolekda) [inline ledger/metrics/src/main/scala/com/daml/metrics/IdentityProviderConfigStoreMetrics.scala:19]: We need to document these new metrics in UNRELEASED.md

File: ledger/participant-integration-api/src/main/scala/platform/LedgerApiServer.scala [hunk 25]
```diff
@@ -147,6 +152,12 @@ class LedgerApiServer(
         cacheExpiryAfterWrite = apiServerConfig.identityProviderManagement.cacheExpiryAfterWrite,
         maxIdentityProviders = IdentityProviderManagementConfig.MaxIdentityProviders,
       )(servicesExecutionContext, loggingContext)
+    val identityProviderConfigLoader = new IdentityProviderConfigLoader {
+      override def getIdentityProviderConfig(issuer: LedgerId)(implicit
+          loggingContext: LoggingContext
+      ): Future[domain.IdentityProviderConfig] =
+        identityProviderStore.getActiveIdentityProviderByIssuer(issuer)
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/LedgerApiServer.scala:159]: which execution context should this run on?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala (hunk 23 above)
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala:103]: So this will actually check that user with this id belongs to the same IDP as the requesting user. If it does not - i.e. there is a user with another "identity_provider_id" - we throw permission denied rather that user exists error. This is the case we discussed before, if user does not exist - this check will pass.

File: ledger/participant-integration-api/src/main/scala/platform/LedgerApiServer.scala (hunk 25 above)
Comment (skisel-da) [reply ledger/participant-integration-api/src/main/scala/platform/LedgerApiServer.scala:159]: servicesExecutionContext ?

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala [hunk 26]
```diff
@@ -362,6 +370,23 @@ private[apiserver] final class ApiPartyManagementService private (
     }
   }
 
+  // Here we check if party exists and actually belongs to the requested Identity Provider
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiPartyManagementService.scala:373]:
```suggestion
  // Check if party either doesn't exist or exists and belongs to the requested Identity Provider
```

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala (hunk 23 above)
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala:103]: Can we then rename this method to better reflect what it does i.e. follow the example of Party Service: verifyUserIsNonExistentOrInIdp

Comment (skisel-da) [reply ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiUserManagementService.scala:103]: sure, Pawel suggested the name for party management service and I forgot to update this one.

File: ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Authorizer.scala (hunk 15 above)
[the remark this answers is not in the document]
Comment (mziolekda) [reply ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/Authorizer.scala:137]: I would include parts of the justification in form of comment. Surely it is not the last time we will be looking at this code asking same questions as Pawel has

**Review by mziolekda — APPROVED** [review]

File: ledger/participant-integration-api/src/main/scala/platform/LedgerApiServer.scala (hunk 25 above)
Comment (mziolekda) [reply ledger/participant-integration-api/src/main/scala/platform/LedgerApiServer.scala:159]: I would think so too
